src/Hint/Restrict.hs

```diff
@@ -105,22 +109,52 @@ checkPragmas modu flags exts mps =
 
 checkImports :: String -> [LImportDecl GhcPs] -> (Bool, Map.Map String RestrictItem) -> [Idea]
 checkImports modu imp (def, mp) =
-    [ ideaMessage riMessage $ if not allowImport
-      then ideaNoTo $ warn' "Avoid restricted module" i i []
-      else warn' "Avoid restricted qualification" i (noLoc $ (unLoc i){ ideclAs=noLoc . mkModuleName <$> listToMaybe riAs} :: Located (ImportDecl GhcPs)) []
+    [ ideaMessage riMessage
+      $ if | not allowImport -> ideaNoTo $ warn' "Avoid restricted module" i i []
+           | not allowIdent  -> ideaNoTo $ warn' "Avoid restricted identifiers" i i []
+           | not allowQual   -> warn' "Avoid restricted qualification" i (noLoc $ (unLoc i){ ideclAs=noLoc . mkModuleName <$> listToMaybe riAs} :: Located (ImportDecl GhcPs)) []
+           | otherwise       -> error "checkImports: unexpected case"
     | i@(LL _ ImportDecl {..}) <- imp
-    , let ri@RestrictItem{..} = Map.findWithDefault (RestrictItem [] [("","") | def] Nothing) (moduleNameString (unLoc ideclName)) mp
+    , let ri@RestrictItem{..} = Map.findWithDefault (RestrictItem [] [("","") | def] [] Nothing) (moduleNameString (unLoc ideclName)) mp
     , let allowImport = within modu "" ri
+    , let allowIdent = Set.disjoint
+                       (Set.fromList riBadIdents)
+                       (Set.fromList (maybe [] (\(b, lxs) -> if b then [] else concatMap (importListToIdents . unLoc) (unLoc lxs)) ideclHiding))
     , let allowQual = maybe True (\x -> null riAs || moduleNameString (unLoc x) `elem` riAs) ideclAs
-    , not allowImport || not allowQual
+    , not allowImport || not allowQual || not allowIdent
     ]
 
+importListToIdents :: IE GhcPs -> [String]
+importListToIdents =
+  catMaybes .
+  \case (IEVar _ n)              -> [fromName n]
+        (IEThingAbs _ n)         -> [fromName n]
+        (IEThingAll _ n)         -> [fromName n]
+        (IEThingWith _ n _ ns _) -> fromName n : map fromName ns
+        IEModuleContents {}      -> error "importListToIdents: this is only supposed to happen in an export list"
+        IEGroup {}               -> []
```

Comment:
Could we just use `_ -> []` for all these? I would rather be less tied to the GHC API if possible.